pulumi/infra/kafka.py

```diff
@@ -1,13 +1,137 @@
+# confluent-cloud-infrastructure project:
+# TODO: link
+KAFKA_SERVICES = [
+    "pipeline-ingress",
+    "graph-generator",
+    "node-identifier",
+    "graph-merger",
+    "analyzer-executor",
+    "engagement-creator",
+]
+
+
+@dataclasses.dataclass
+class CredentialOutput:
+    service_account_id: str
+    api_key: str
+    api_secret: str
+
+    @staticmethod
+    def from_json(json_: Mapping[str, str]) -> CredentialOutput:
+        return CredentialOutput(
+            service_account_id=json_["service_account_id"],
+            api_key=json_["api_key"],
+            api_secret=json_["api_secret"],
+        )
+
+
+@dataclasses.dataclass
+class TopicOutput:
+    partitions: int
+    config: Mapping[str, Any]
+
+    @staticmethod
+    def from_json(json_: Mapping[str, Any]) -> TopicOutput:
+        return TopicOutput(
+            partitions=json_["partitions"],
+            config=json_["config"],
+        )
+
+
+@dataclasses.dataclass
+class ServiceOutput:
+    ingress_topics: Sequence[str]
+    egress_topics: Sequence[str]
+    service_account: CredentialOutput
+
+    @staticmethod
+    def from_json(json_: Mapping[str, Any]) -> ServiceOutput:
+        return ServiceOutput(
+            ingress_topics=json_["ingress_topics"],
+            egress_topics=json_["egress_topics"],
+            service_account=CredentialOutput.from_json(json_["service_account"]),
+        )
+
+
+@dataclasses.dataclass
+class EnvironmentOutput:
+    environment_id: str
+    bootstrap_servers: str
+    environment_credentials: CredentialOutput
+    services: Mapping[str, ServiceOutput]
+    topics: Mapping[str, TopicOutput]
+
+    def get_service_credentials(self, service_name: str) -> CredentialOutput:
+        if service_name in self.services:
+            return self.services[service_name].service_account
+        else:
+            raise KeyError(f"{service_name} does not exist")
+
+    @staticmethod
+    def from_json(json_: Mapping[str, Any]) -> EnvironmentOutput:
+        return EnvironmentOutput(
+            environment_id=json_["environment_id"],
+            bootstrap_servers=json_["bootstrap_servers"],
+            environment_credentials=CredentialOutput.from_json(
+                json_["environment_credentials"]
+            ),
+            services={
+                k: ServiceOutput.from_json(v) for k, v in json_["services"].items()
+            },
+            topics={k: TopicOutput.from_json(v) for k, v in json_["topics"].items()},
+        )
+
+
+@dataclasses.dataclass
+class ConfluentOutput:
+    environments: Mapping[str, EnvironmentOutput]
+
+    def get_environment(self, environment_name: str) -> EnvironmentOutput:
+        if environment_name in self.environments:
+            return self.environments[environment_name]
+        else:
+            raise KeyError(f"{environment_name} does not exist")
+
+    @staticmethod
+    def from_json(
+        json_: pulumi.Output[Mapping[str, Any]]
```

Comment:
Why is this a `pulumi.Output[Mapping[...]]`, but all the other `from_json` methods take just a `Mapping[...]`?

Reply:
This is the outermost layer of a nested structure, so only this method receives an actual `Output[T]`. All the others are operating inside an `.apply(..)`.

Reply:
Ah... thought it might be something like that.

Might be worth adding a comment to that effect somewhere in this file.